Content.Server/GameObjects/EntitySystems/Weapon/Melee/MeleeWeaponSystem.cs

```diff
@@ -0,0 +1,267 @@
+
+            for (var i = 0; i < increments; i++)
+            {
+                var castAngle = new Angle(baseAngle + increment * i);
+                var res = EntitySystem.Get<SharedBroadPhaseSystem>().IntersectRay(mapId,
+                    new CollisionRay(position, castAngle.ToWorldVec(),
+                        (int) (CollisionGroup.Impassable | CollisionGroup.MobImpassable)), range, ignore).ToList();
+
+                if (res.Count != 0)
+                {
+                    resSet.Add(res[0].HitEntity);
+                }
+            }
+
+            return resSet;
+        }
+
+        private void OnChemicalInjectorHit(EntityUid uid, MeleeChemicalInjectorComponent comp, MeleeHitEvent args)
+        {
+            if (!ComponentManager.TryGetComponent<SolutionContainerComponent>(uid, out var solutionContainer))
+                return;
+
+            var hitBloodstreams = new List<BloodstreamComponent>();
+            foreach (var entity in args.HitEntities)
+            {
+                if (entity.Deleted)
+                    continue;
+
+                if (entity.TryGetComponent<BloodstreamComponent>(out var bloodstream))
+                    hitBloodstreams.Add(bloodstream);
+            }
+
+            if (!hitBloodstreams.Any())
+                return;
+
+            var removedSolution = solutionContainer.Solution.SplitSolution(comp.TransferAmount * hitBloodstreams.Count);
+            var removedVol = removedSolution.TotalVolume;
+            var solutionToInject = removedSolution.SplitSolution(removedVol * comp.TransferEfficiency);
+            var volPerBloodstream = solutionToInject.TotalVolume * (1 / hitBloodstreams.Count);
+
+            foreach (var bloodstream in hitBloodstreams)
+            {
+                var individualInjection = solutionToInject.SplitSolution(volPerBloodstream);
+                bloodstream.TryTransferSolution(individualInjection);
+            }
+        }
+
+        private void OnCooldownRefreshed(EntityUid uid, ItemCooldownComponent comp, RefreshItemCooldownEvent args)
+        {
+            comp.CooldownStart = args.LastAttackTime;
+            comp.CooldownEnd = args.CooldownEnd;
+        }
+
+        public void SendAnimation(string arc, Angle angle, IEntity attacker, IEntity source, IEnumerable<IEntity> hits, bool textureEffect = false, bool arcFollowAttacker = true)
+        {
+            RaiseNetworkEvent(new MeleeWeaponSystemMessages.PlayMeleeWeaponAnimationMessage(arc, angle, attacker.Uid, source.Uid,
+                hits.Select(e => e.Uid).ToList(), textureEffect, arcFollowAttacker), Filter.Pvs(source, 1f));
+        }
+
+        public void SendLunge(Angle angle, IEntity source)
+        {
+            RaiseNetworkEvent(new MeleeWeaponSystemMessages.PlayLungeAnimationMessage(angle, source.Uid), Filter.Pvs(source, 1f));
+        }
+    }
+
+    public class MeleeHitEvent : EntityEventArgs
+    {
+        public readonly List<IEntity> HitEntities;
```

Comment:
Change this for something like `IEnumerable<IEntity>` or `IReadOnlyList<IEntity>`, so it can't be mutated. Lists cast to both of those cleanly.